pinot-tools/src/main/java/org/apache/pinot/tools/PinotDataAndQueryAnonymizer.java

```diff
@@ -0,0 +1,1287 @@
+            }
+          };
+        case STRING:
+          _comparator = new Comparator() {
+            @Override
+            public int compare(Object o1, Object o2) {
+              return ((String)o1).compareTo((String) o2);
+            }
+          };
+          break;
+        default:
+          throw new UnsupportedOperationException("global dictionary currently does not support: " + _dataType.name());
+      }
+    }
+
+    void sort() {
+      Arrays.sort(_origValues, 0, _index, _comparator);
+    }
+
+    void addOrigValue(Object origValue) {
+      if (!linearSearch(origValue)) {
+        _origValues[_index++] = origValue;
+      }
+    }
+
+    void setDerivedValues(Object[] derivedValues) {
+      Preconditions.checkState(derivedValues.length == _index);
+      _derivedValues = derivedValues;
+    }
+
+    Object getDerivedValueForOrigValue(Object origValue) {
+      int index = binarySearch(0, _index - 1, origValue);
+      Preconditions.checkState(index >= 0, "Expecting origValue: " + origValue);
+      return _derivedValues[index];
+    }
+
+    // used during building global dictionary phase
+    // to prevent adding original values already
+    // seen from previous segments
+    boolean linearSearch(Object key) {
+      for (int i = 0; i < _index; i++) {
+        if (key == null) {
+          if (_origValues[i] == null) {
+            return true;
+          }
+        } else {
+          if (_origValues[i].equals(key)) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+
+    // used during data generation phase.
+    // since the global dictionary is fully built
+    // and sorted, we can do binary search
+    int binarySearch(int low, int high, Object key) {
+      if (low > high) return -1;
+
+      int mid = (low + high)/2;
+
+      if (_origValues[mid].equals(key)) {
+        return mid;
+      }
+
+      if (isLessThan(_origValues[mid], key) < 0) {
+        return binarySearch(mid + 1, high, key);
+      }
+
+      return binarySearch(low, mid - 1, key);
+    }
+
+    private int isLessThan(Object o1, Object o2) {
+      switch (_dataType) {
+        case INT:
+          return ((Integer) o1).compareTo((Integer)o2);
+        case LONG:
+          return ((Long) o1).compareTo((Long)o2);
+        case DOUBLE:
+          return ((Double) o1).compareTo((Double) o2);
+        case FLOAT:
+          return ((Float) o1).compareTo((Float) o2);
+        case STRING:
+          return ((String) o1).compareTo((String) o2);
+        default:
+          throw new IllegalStateException("unexpected data type: " + _dataType);
+      }
+    }
+  }
+
+  public void buildGlobalDictionaries() throws Exception {
+    if (_globalDictionaryColumns.isEmpty()) {
+      LOGGER.info("Set of global dictionary columns is empty");
+      return;
+    }
+
+    _timeToBuildDictionaries.start();
+
+    File segmentParentDirectory = new File(_segmentDir);
```

Comment:
Will be useful to do the lines 406 through 426 on a per column basis, so that we don't have to hold all values of all columns in memory. Feel free to add a TODO for optimization

Reply:
yes we can do that. I added it to the list of items for "planned work" at the beginning of file. During data generation we load the global dictionaries, we have to load for all the columns or we can mmap it.